Rank Widget

grzesiek2010 opened the pull request:
Closes #33 

#### What has been done to verify that this works as intended?
Tested using forms:
[rank.xml.txt](https://github.com/opendatakit/collect/files/2128940/rank.xml.txt)
[filter-rank.xml.txt](https://github.com/opendatakit/collect/files/2128941/filter-rank.xml.txt)

#### Why is this the best possible solution? Were any other approaches considered?
It uses a common implementation using RecyclerView so it loos and works well and I'm not able to imagine a better solution.

#### Are there any risks to merging this code? If so, what are they?
It's a new feature and shouldn't affect any other things so I would say it's safe.

#### Do we need any specific form for testing your changes? If so, please attach one.
Forms attached above.

#### Does this change require updates to documentation? If so, please file an issue at [Here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
https://github.com/opendatakit/docs/issues/747

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lint findbugs` and confirmed all checks still pass.
- [x] verified that any new UI elements use theme colors so that they work with both light and dark themes.

grzesiek2010 committed:
RankWidget

xsteelej commented:
Hi @grzesiek2010, looks like the build + tests failed

grzesiek2010 commented:
It's just for testing since there is a related pr on the Javarosa's side we should merge first so I don't really care about the build now. It's even marked as WIP as you see and the description is not filled.

xsteelej commented:
Hi @grzesiek2010, ok that explains it.. I missed the "WIP:" in the title as meaning Work In Progress as it was marked as Review Required.

lognaturel commented:
I gave this a try with [rank.xml.txt](https://github.com/opendatakit/collect/files/2007094/rank.xml.txt). It's a great feature, nice work, @grzesiek2010!

I'm seeing the value rather than the label show up for each item. Could it be an issue with `jr:itext`?

I'm also wondering about the delay for "picking" an item. It took me a while to figure that out since it's not standard. Is it necessary because of the swiping? For usability, I think it would be better to reduce the delay or remove it.

There hopefully will be more feedback coming at https://forum.opendatakit.org/t/collect-question-type-rank-items/12568/6

lognaturel commented:
The forum conversation at https://forum.opendatakit.org/t/collect-question-type-rank-items/12568 has concluded and ranking subsets will be done by using a choice filter. That means no big changes needed here!

@grzesiek2010 I tried subset ranking with [filter-rank.xml.txt](https://github.com/opendatakit/collect/files/2018407/filter-rank.xml.txt) and it works well with one small issue -- if a user goes back to the select multiple and selects more options than were previously chosen, the rank widget doesn't show those new options.

I also noticed that when choices are dragged outside the bounds of the list, then disappear which is kind of confusing. I'm not sure if there's a great alternative but wanted to point it out in case you have good ideas.

grzesiek2010 commented:
> f a user goes back to the select multiple and selects more options than were previously chosen, the rank widget doesn't show those new options.

what should happen in that case? For example:
1. You select 3 items and go forward
2. Change the order of selected elements and go back
3. Add one another item

Should I keep the saved order and add a new item at the bottom or maybe I should use the default order after changing the set of items?

grzesiek2010 committed:
Add separators between rank elemsnts too

grzesiek2010 committed:
Prevent out of bounds dragging

grzesiek2010 committed:
Add rest of items if you selected new that should be also ordered

grzesiek2010 commented:
> I'm seeing the value rather than the label show up for each item.

Fixed.

> I also noticed that when choices are dragged outside the bounds of the list, then disappear which is kind of confusing.

Fixed.

> if a user goes back to the select multiple and selects more options than were previously chosen, the rank widget doesn't show those new options

I fixed that adding new items at the bottom of the list let me know if it's ok (the question above).

> I'm also wondering about the delay for "picking" an item. It took me a while to figure that out since it's not standard. Is it necessary because of the swiping?

I think dragging onTouch without delaying is used if it's triggered after touching on a small area of a view like [drag icon](https://material.io/tools/icons/?search=drag&icon=drag_handle&style=baseline), you said we don't want to add that icon so making the entire view draggable on touch is not a good idea since it will cause scrolling not possible.

grzesiek2010 committed:
Adjusted colors to light/dark themes

lognaturel commented:
> adding new items at the bottom of the list

Looks good to me, thanks! The behavior when reaching the top and bottom of the list is now great, too.

I'd like another opinion on the dragging behavior. @jd-alexander, @smoyth, @yanokwa, anyone else? Currently we need to long press to move an item. Even I seem to forget to long press so maybe a drag handle would be a helpful visual cue after all.

> making the entire view draggable on touch is not a good idea since it will cause scrolling not possible

That makes sense.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/adapters/RankingListAdapter.java:
```diff
@@ -0,0 +1,98 @@
+
+public class RankingListAdapter extends RecyclerView.Adapter<RankingListAdapter.ItemViewHolder> {
+
+    private Context context;
```
I don't think you need this context field. It's only used in `onCreateViewHolder` but you could use `parent.getContext()` there I believe.

grzesiek2010 replied:
Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/adapters/RankingListAdapter.java:
```diff
@@ -0,0 +1,98 @@
+        holder.textView.setText(formEntryPrompt.getSelectChoiceText(items.get(position)));
+    }
+
+    public boolean onItemMove(int fromPosition, int toPosition) {
```
Do you really need to return here?

grzesiek2010 replied:
You are right, Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/adapters/RankingListAdapter.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
```
Annotate with `@NonNull`

grzesiek2010 replied:
Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/ThemeUtils.java:
```diff
@@ -108,4 +108,12 @@ public int getAccentColor() {
+     */
+    @ColorInt
+    public int getRankItemColor() {
+        return getAttributeValue(R.attr.rankItemColor);
```
This could be something more generic like "primary background color" or something like that.

grzesiek2010 replied:
Fixed.

jd-alexander commented:
Yes I agree that long presses for drag might not be ideal and the option of having the drag handle should work well to trigger the drag functionality. I also looked at the Material Design icon for drags and honestly it doesn't seem to communicate dragging/movement so I went and did some research and found a [great discussion on Stack Exchange](https://ux.stackexchange.com/questions/34158/how-to-make-it-obvious-that-you-can-drag-things-that-you-normally-cant) about this, where they used a four way arrow to denote dragging.

grzesiek2010 committed:
Fixed bugs

grzesiek2010 commented:
> they used a four way arrow

on the other hand, that icon might indicate that we allow moving in four directions.

The problem is not only how that icon might look like but also that if we use any icon then we don't have space for media files (audio/image/video) - they are available in SlectOne/SelectMultiple so we may want to add them for RankWidget also in the future.

lognaturel commented:
@grzesiek2010 Good point about media. It should just work as-is, right? Could you please add a question with media ranking to the test form so we make sure to verify it?

How about getting the current behavior out to users in a beta to get some feedback?

One other small thing I noticed is that "Remove response" currently has no effect. Should it perhaps reset the order to the original one?

grzesiek2010 commented:
> Good point about media. It should just work as-is, right?

It won't work out of the box since `MediaLayout` is not used in `RankingWidget` it would require some changes so I think it's a good improvement we can add in a separate pr.

> One other small thing I noticed is that "Remove response" currently has no effect. Should it perhaps reset the order to the original one?

I'll look at it. Don't you expect supporting null values as it was in `RangeWidget`? Is reverting an original order ok?

grzesiek2010 committed:
Fixed removeResponse option

grzesiek2010 commented:
> One other small thing I noticed is that "Remove response" currently has no effect. Should it perhaps reset the order to the original one?

Fixed.

lognaturel commented:
Sigh. Martijn points out that really it should be possible to represent an empty or unsorted state. See https://forum.opendatakit.org/t/collect-question-type-rank-items/12568/19. Does a button to "Start ranking" seem ok? Any other UI ideas?

grzesiek2010 committed:
Support null values

lognaturel commented:
👏 Thanks, @grzesiek2010! I do like that this matches what is planned for Enketo. Since we know we're on mobile, the text should say "Tap" rather than "Click".

One possible issue is that if the list gets long the instruction text might not fit. Probably ranking only makes sense with relatively short lists, though?

I think this is ready to test and go to beta. We can get feedback from the broader community and then make any tweaks needed. @grzesiek2010 can you please fill out the PR template when you get a chance?

lognaturel commented:
Hmm, here's a deeper problem with the null value interface: swiping on the disabled choices enables the widget and gives the question a value. The only way to leave the question blank is to swipe on the question label. The options get enabled late in the swipe so it's not very clear that's happening.

I think we should go with a consistent experience across widgets. That is, the null case would be represented with a button only as in with dates, geoshapes, etc. Once the button is tapped, the ranking interface would be shown and reflect the order in the answer. If the user cleared the answer, it would go back to only showing the button. What do others think?

grzesiek2010 commented:
>  think we should go with a consistent experience across widgets. That is, the null case would be represented with a button only as in with dates, geoshapes, etc. Once the button is tapped, the ranking interface would be shown and reflect the order in the answer. If the user cleared the answer, it would go back to only showing the button. What do others think?

not a beautiful UI but better behavior so generally I agree but I think we should do that in a separate pr since the same you want to do with `RangeWidget` right?

grzesiek2010 commented:
> the text should say "Tap" rather than "Click"

Fixed.

grzesiek2010 committed:
Start ordering by double tap

grzesiek2010 commented:
@lognaturel what about using double tap to start? I like this solution please check my last changes.

lognaturel commented:
Good idea! Let'd do a beta with that and get some feedback.

mmarciniak90 commented:
@grzesiek2010
* I used form filter-rank.xml and noted one case
![rank](https://user-images.githubusercontent.com/3907554/41649705-bb315c1a-746b-11e8-91b8-cd643f1a5671.gif)
As you can see the options are loaded one by one not all at once.

* Aggregate does not accept the rank form
![aggregate](https://user-images.githubusercontent.com/3907554/41649864-22ac23e8-746c-11e8-8981-96806f53a40a.png)

* If it is not a big deal, it is worth adding the  warning about spaces in underlying values

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

grzesiek2010 committed:
Fixed bug with new added items - all should be displayed

grzesiek2010 committed:
Display a warning about values with spaces if needed

grzesiek2010 committed:
Don't display 'Double tap to start' message if an itemset is empty

lognaturel commented:
I'm going to rebase on master to get the latest JR snapshot and then merge this so we can more easily issue a beta and get user feedback. If any other problems are found, they can be filed as issues.

grzesiek2010 committed:
Fixed pmd and checkstyle errors
